src/err/mod.rs

```diff
@@ -239,10 +244,18 @@ impl PyErr {
                 std::panic::resume_unwind(Box::new(msg))
             }
 
-            err
+            Some(err)
         }
     }
 
+    /// Retrieves the current error from the Python interpreter's global state.
+    ///
+    /// The error is cleared from the Python interpreter.
+    /// Panics if no error is set
```

Comment:
This behavior is new compared to before, where it returned a `SystemError`.

Should we deviate from the Python convention here?

Reply:
I'm not sure, see https://github.com/PyO3/pyo3/pull/1715#issuecomment-873612388

Reply:
I mean, I fully agree with `PyErr::fetch` returning an option. But the thing that we used for our own calls, now called `api_call_failed` used to return a SystemError and now panics.

Reply:
We're not sure if it really needs to return a `SystemError`, so I'm trying to just panic now to see if anything breaks.

Reply:
@birkenfeld what do you mean by "the Python convention" above?

I'm personally going back and forth on whether this should panic. On one hand, it feels like if this ever happens it's a genuine bug - we think a Python error has occurred but we found no exception when we went to retrieve it.

On the other hand, we translate panics into exceptions which are pretty hard to catch for Python users, so it might be nicer to continue to use `SystemError` anyway.

Reply:
When a CPython function returns NULL/-1/... but no exception is set (i.e. someone wrote a buggy C function), the eval loop raises a `SystemError: error return with no exception set`.  This is equivalent to the case where `api_call_failed` is used.

Reply:
Aha yes, found it: https://github.com/python/cpython/blob/17f94e28882e1e2b331ace93f42e8615383dee59/Python/ceval.c#L4330-L4333

In which case I would definitely vote for us returning a `SystemError` with the same error message as CPython!

Reply:
Looks like CPython only raises a `SystemError` in release build, do we want to follow that too and only raises a `SystemError` in release mode and panic in debug mode?

Reply:
I don't mind. Having the panic message printed would probably help debugging, because it'll include the Rust traceback. The Python `SystemError` will only have a traceback which starts at the point where Python called Rust.

@birkenfeld what do you think on that option?

Reply:
Yeah it's fine.  This "should not happen" anyway, and if it does it's as likely to be a bug in the Rust code due to misunderstanding the error condition.